### PR TITLE
Logs: use explicit frame-type-check

### PR DESCRIPTION
**Pull request by gabor**
the dataframes returned by Loki for logs contain a special field named "labels".
in the generic logs-code, previously we just looked for a field named "labels", and if it was there, we assumed it's the thing with the labels in it. but there can be false positives, an another datasource might return data with a `labels` field that is not related to this.

to be more explicit, we should create a new frame-type value for this format. that's happening at https://github.com/grafana/grafana-plugin-sdk-go/pull/483 .

until that PR is merged, we will use a custom-field in the frame's meta to say that we have this format. this will fix the false-positive problem.

how to test:
1. make sure you have the `lokiBackendMode` feature-flag enabled
2. have a loki datasource, go to grafana-explore, run a loki logs query that returns data with labels
3. in the logs-visualization, open a log-line, and verify that there is a "Log labels" section, with content in it, not just a "Detected fields" section
4. now disable the `lokiBackendMOde` feature-flag, and repeat the test.

### CODE REVIEW

File: public/app/core/logs_model.ts [hunk 1]
```diff
@@ -329,10 +329,11 @@ function getLabelsForFrameRow(fields: LogFields, index: number): Labels {
...
 
-  return {
-    ...stringField.labels,
-    ...labelsField?.values.get(index),
-  };
+  if (labelsField !== undefined) {
+    return labelsField.values.get(index);
+  } else {
+    return stringField.labels ?? {};
+  }
```
Comment (gabor) [inline public/app/core/logs_model.ts:336]: i made the code stricter. if your data has the labels-field (which gets found only if you have the specific-frame-type), we only use that source of information

File: public/app/plugins/datasource/loki/backendResultTransformer.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ function processStreamFrame(
     limit: query?.maxLines,
     searchWords: query !== undefined ? getHighlighterExpressionsFromQuery(formatQuery(query.expr)) : undefined,
     custom: {
+      ...frame.meta?.custom, // keep the original meta.custom
```
Comment (gabor) [inline public/app/plugins/datasource/loki/backendResultTransformer.ts:33]: i had a bug here where we just overwrote the `frame.meta.custom`  that was coming from the server.

**Review by joey-grafana — COMMENTED** [review]
I get the same UI whether or not I have lokiBackendMode enabled/disabled: that might be expected but the `[object Object]` is probably a problem? (I think it's been there for a while)

Screenshot 2022-05-15 at 18 20 51

**Comment by gabor** [comment]
@joey-grafana 
> I get the same UI whether or not I have lokiBackendMode enabled/disabled: that might be expected but the `[object Object]` is probably a problem? (I think it's been there for a while)

thanks for the review!

the `[object Object]` thing is something i'd prefer to fix in a separate PR if that's OK with you
( basically, we need to adjust this IF-check: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Logs/logParser.ts#L68-L71 with something like"if frametype-matches and it's called `labels` then ignore it", but that file is in `grafana-ui`, where whatever i put in there, it needs to be backward-compatible, so first i'm waiting for this new frame-type to be "official", and then i'll put it there :-) )

**Review by joey-grafana — APPROVED** [review]

**Pull request merged**